### DISCUSSION TITLE
Is `Plot.barX(data, { y: (d) => makeY(d), ...})` equivalent to `Plot.barX(data, { y: { value: (d) => makeY(d) }, ...})` ?

### DISCUSSION THREAD

**Discussion by guizmaii** [observablehq/plot#1869] [Q&A] [answered]
Hi everyone 🙂

I'm struggling a bit with the facets.
I have a bar chart which is grouped thanks to facets, sadly the `Plot.text` that I'm adding to my chart to add show the data of each bar isn't working well (I didn't find how to make work yet). 
Instead of each bar being associated with its data, I have each bar associated with many data.
I'm expecting this:
```
bar a - | ========== <value a0>         | 
bar b - | ============== <value b0>     | facet 0 
bar c - | ====== <value c0>             | 

bar a - | ============== <value a1>     | 
bar b - | ============ <value b1>       | facet 1 
bar c - | ======== <value c1>           | 
```

But I have this:
```
bar a - | ========== <value a0 mixed with value a1>         | 
bar b - | ============== <value b0 mixed with value b1>     | facet 0 
bar c - | ====== <value c0 mixed with value c1>             | 

bar a - | ============== <value a0 mixed with value a1>     | 
bar b - | ============ <value b0 mixed with value b1>       | facet 1 
bar c - | ======== <value c0 mixed with value c1>           | 
```

While trying to make this work, I tried to change my `y` definition in my `Plot.barX` call.
When I use:
```js
Plot.barX(data, { y: (d) => makeY(d), ...})
``` 
my bars are showing up but when I use:
```js
Plot.barX(data, { y: { value: (d) => makeY(d) }, ...})
```

Seeing the type signature of the `y` parameter:
<img width="700" alt="Screenshot 2023-09-22 at 6 43 46 pm" src="https://github.com/observablehq/plot/assets/1193670/7a3499b2-4450-4895-b98e-c330389f72a0">
<img width="877" alt="Screenshot 2023-09-22 at 6 43 55 pm" src="https://github.com/observablehq/plot/assets/1193670/ad2112a0-23af-48a3-b42b-f25509a98fb9">

I was expecting these two pieces of code to be equivalent but the results aren't, why? 🤔

**Comment by mbostock** [comment, marked as answer]
As to the intent, I don’t follow the problem you describe, but based on the ASCII art that you shared here’s how I might produce a horizontal bar chart with a value label for each bar:

![untitled (91)](https://github.com/observablehq/plot/assets/230541/9949fed0-e55c-446d-85ef-63bb9d5d9d39)

```js
Plot.plot({
  x: {insetRight: 40, grid: true},
  marks: [
    Plot.barX(data, {y: "name", x: "value", fy: "facet", fill: "steelblue"}),
    Plot.ruleX([0]),
    Plot.text(data, {text: "value", y: "name", x: "value", fy: "facet", frameAnchor: "left", dx: 6}),
    Plot.frame(),
  ]
})
```
 
The data looks like this (CSV with _name_, _value_, and _facet_ columns):

```js
data = [
  {name: "a", value: 10, facet: 0},
  {name: "b", value: 14, facet: 0},
  {name: "c", value: 6, facet: 0},
  {name: "a", value: 14, facet: 1},
  {name: "b", value: 12, facet: 1},
  {name: "c", value: 8, facet: 1},
]
```

Notebook: https://observablehq.com/d/f671d7a067574995

Is that what you want? If not, can you share some live code that demonstrates the problem, ideally as a notebook?

As to the technical question in the title: this is a bug with a few transforms in Plot, including the stack transform used implicitly by the bar mark. These transforms don’t currently check for a ChannelValueSpec and assume that a ChannelValue is being passed directly. This is discussed a little in #1364. I filed #1873 to track this more explicitly.

You can use a ChannelValueSpec with a mark directly, and hence it is supported with Plot.barX. But you have to be careful to avoid the implicit stack transform. So this:

![untitled (89)](https://github.com/observablehq/plot/assets/230541/35cd693f-438e-451c-b186-790691eab1a3)

```js
Plot.plot({
  marks: [
    Plot.barX(data, {y: (d) => d.name, x: (d) => d.value})
  ]
})
```

Can be rewritten as:

![untitled (90)](https://github.com/observablehq/plot/assets/230541/46d17bee-0b2d-4807-aaf8-63d85089013a)

```js
Plot.plot({
  marks: [
    Plot.barX(data, {y: {value: (d) => d.name}, x2: (d) => d.value})
  ]
})
```

But note the use of **x2** rather than **x** to avoid the implicit stack transform.

Notebook: https://observablehq.com/d/2cac59eacd3ce96b

The problem originates here in the stack transform, where we call Plot.valueof directly with the options values, rather than distinguishing ChannelValueSpec from ChannelValue:

https://github.com/observablehq/plot/blob/1b1da9f2535371fb08f8116bc28f6ef78fc44f63/src/transforms/stack.js#L90-L92

@Fil discusses making this change to Plot.valueof in #1855, saying:

> But this is such an encompassing change, that I'm a bit nervous, and confused to know if it's correct.

So ChannelValueSpec is well supported by marks, but the support by transforms is a bit dicey and needs to be improved.

**Reply by guizmaii** [reply]
Thanks for your very detailed answer, Mike 🙂

I found the root cause of my issue. Dumb mistake as always 🤦‍♂️
Forgot to add the `fy: ...` parameter on my `Plot.text`. It was only on my `Plot.barX`
